### PR TITLE
Remove ServerView from RealtimeIndexTasks and use coordinator http endpoint for handoff information

### LINKED ISSUE #1970: Peons put high load on zookeeper on disconnects
Realtime Index tasks need to know when the segments is handed over to the historical nodes. 
To get this information realtime index task keeps a FilteredBrokerserverView and gets data from zookeeper for all the segments in the cluster. In case of ZK disconnects, each task will reconnect to zookeeper and starts reading the distribution of segments in the cluster which puts a high load on zookeeper as the number of tasks starts to grow. 

One solution to handle this is to make the overlord responsible for keeping the state segments in cluster and the realtime nodes will interact with the overlord to know when a segment has been handed off to another historical node.

### PR DESCRIPTION
**Pull request by nishantmonu51**
- fixes #1970
- extracted out segment handoff callbacks in SegmentHandoffNotifier
which is responsible for tracking segment handoffs and doing callbacks
when handoff is complete.
- Coordinator now maintains a timeline for the segments which makes it faster to provide serverView for an interval
- Added new end point to DatasourcesResource which exposes information about where nodes segments are loaded based on interval. 
- realtime index task  and realtime nodes now use HTTP end points exposed by the
coordinator to get serverView

Below image shows the improvements by above changes on zookeeper read load with nearly 500 realtime index tasks running - 

screen shot 2015-12-09 at 3 21 58 am

### CODE REVIEW

File: common/src/main/java/io/druid/timeline/TimelineLookup.java [hunk 1]
```diff
@@ -35,6 +35,8 @@
    */
   public Iterable<TimelineObjectHolder<VersionType, ObjectType>> lookup(Interval interval);
 
+  public Iterable<TimelineObjectHolder<VersionType, ObjectType>> lookup(Interval interval, boolean incompleteOk);
```
Comment (xvrl) [inline common/src/main/java/io/druid/timeline/TimelineLookup.java]: `allowIncomplete` would be clearer I think. Maybe add some javadoc as well while we're at it.

File: indexing-service/src/main/java/io/druid/indexing/common/TaskToolboxFactory.java [hunk 2]
```diff
@@ -75,7 +76,6 @@ public TaskToolboxFactory(
       DataSegmentMover dataSegmentMover,
       DataSegmentArchiver dataSegmentArchiver,
       DataSegmentAnnouncer segmentAnnouncer,
-      FilteredServerView newSegmentServerView,
```
Comment (xvrl) [inline indexing-service/src/main/java/io/druid/indexing/common/TaskToolboxFactory.java]: minor nit, but we should keep the arguments in the same order and just replace newSegmentServerView with handoffNotifierFactory. Makes it easier to track changes and resolve conflicts.

File: server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+
+  static boolean isHandOffComplete(List<SegmentLoadInfo> serverView, SegmentDescriptor descriptor)
+  {
+    for (SegmentLoadInfo segmentLoadInfo : serverView) {
+      if (segmentLoadInfo.getSegment().getInterval().contains(descriptor.getInterval())
+          && segmentLoadInfo.getSegment().getShardSpec().getPartitionNum()
+             == descriptor.getPartitionNumber()
+          && segmentLoadInfo.getSegment().getVersion().compareTo(descriptor.getVersion()) >= 0
```
Comment (xvrl) [inline server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java]: Does this apply the same logic as we currently do? If so, that's fine.

File: server/src/main/java/io/druid/client/CoordinatorServerView.java [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+import io.druid.timeline.partition.PartitionChunk;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+
+/**
+ * ServerView of coordinator for the state of segments being loaded in the cluster.
+ */
+public class CoordinatorServerView implements TimelineInventoryView
```
Comment (cheddar) [inline server/src/main/java/io/druid/client/CoordinatorServerView.java]: Why does this implement `TimelineInventoryView`?  I might not have read the code fully yet, but it seems like it would only expose TimelineInventoryView if it is hoping to expose things based on overshadowing (i.e. the "currently active" segments).  However, in my view, the coordinator would just be exposing information about the cluster, not necessarily only the currently active segments.

Rather, it seems like this should be exposing just a server-based view of all segments that exist and maybe also a segment-based view of all servers that have said segment.

File: server/src/main/java/io/druid/client/coordinator/CoordinatorClient.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+  private final HttpClient client;
+  private final ObjectMapper jsonMapper;
+  private final ServerDiscoverySelector selector;
+
+  @Inject
+  public CoordinatorClient(
+      @Global HttpClient client,
+      ObjectMapper jsonMapper,
+      @Coordinator ServerDiscoverySelector selector
```
Comment (cheddar) [inline server/src/main/java/io/druid/client/coordinator/CoordinatorClient.java]: Using this ServerDiscoverySelector is going to further in-grain having two methods of service discovery for Druid: the internal one using the `Announcer` and Curator's ServiceDiscovery stuff.  Can we instead have the coordinators start announcing their existence using the `Announcer` and watch for them with that mechanism instead of Curator's Service Discovery?  This is technically an internal service discovery use case, so it would be great to use the internal mechanisms (the Curator one is intended for exposure to external services, not internal to the cluster).

Comment (cheddar) [inline server/src/main/java/io/druid/client/coordinator/CoordinatorClient.java]: Actually, now that I think about it, there is a thing that the coordinators already use to know who the leader is.  It would be nice to just latch on to that so that we are always querying the leader.

File: server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    log.info("Adding SegmentHandoffCallback for dataSource[%s] Segment[%s]", dataSource, descriptor);
+    handOffCallbacks.put(descriptor, new Pair<>(exec, handOffRunnable));
+  }
+
+  @Override
+  public void start()
+  {
+    scheduledExecutor = Execs.scheduledSingleThreaded("coordinator_handoff_scheduled_%d");
+    scheduledExecutor.scheduleAtFixedRate(
+        new Runnable()
```
Comment (cheddar) [inline server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java]: Always make sure to wrap the entire `Runnable` that you are scheduling in a try/catch if you want it to actually continue running.  If an exception *ever* makes it out of the `Runnable`, the schedule will halt.  You have a try/catch over some of the business logic, but what happens if you get an `NPE` on 80 or for some reason `itr.next()` throws an exception or something.

File: server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+          @Override
+          public void run()
+          {
+            Iterator<Map.Entry<SegmentDescriptor, Pair<Executor, Runnable>>> itr = handOffCallbacks.entrySet()
+                                                                                                   .iterator();
+            while (itr.hasNext()) {
+              Map.Entry<SegmentDescriptor, Pair<Executor, Runnable>> entry = itr.next();
+              SegmentDescriptor descriptor = entry.getKey();
+              try {
+                List<SegmentLoadInfo> serverView = coordinatorClient.fetchServerView(
```
Comment (cheddar) [inline server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java]: I believe this is representing the current set of segments and who is serving them.  In that case `serverView` doesn't seem like the best name to me.  How about `loadedSegments`?

File: server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+                    true
+                );
+
+                if (isHandOffComplete(serverView, entry.getKey())) {
+                  log.info("Segment Handoff complete for dataSource[%s] Segment[%s]", dataSource, descriptor);
+                  entry.getValue().lhs.execute(entry.getValue().rhs);
+                  itr.remove();
+                } else {
+                  log.info(
+                      "Segment Handoff not complete for dataSource[%s] Segment[%s]. Will try again after [%d]secs",
```
Comment (cheddar) [inline server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java]: If you add a `,` to `%d` the millis will be easier to read.  I.e.

```
Will try again after [%,d]secs"
```

File: server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java [hunk 9]
```diff
@@ -597,6 +584,17 @@ public void doRun()
           }
         }
     );
+    handoffNotifier.registerSegmentHandoffCallback(
```
Comment (cheddar) [inline server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java]: For some reason the diff on this file came out really difficult to read.  Can you provide a high level statement of what you changed?

File: server/src/main/java/io/druid/client/coordinator/CoordinatorClient.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    try {
+      StatusResponseHolder response = client.go(
+          new Request(
+              HttpMethod.GET,
+              new URL(
+                  String.format(
+                      "%s/datasources/%s/intervals/%s/serverview?partial=%s",
+                      baseUrl(),
+                      dataSource,
+                      URLEncoder.encode(interval.toString(), "UTF-8"),
```
Comment (cheddar) [inline server/src/main/java/io/druid/client/coordinator/CoordinatorClient.java]: Should we convert the `/` to a `_`?  It looks like you are URL Encoding it, which should also work, but just replacing it with an underscore might be simpler?

File: common/src/main/java/io/druid/timeline/TimelineLookup.java (hunk 1 above)
Comment (cheddar) [inline common/src/main/java/io/druid/timeline/TimelineLookup.java]: Can we instead expose the method

```
public Iterable<TimelineObjectHolder<VersionType, ObjectType>> lookupWithIncomplete(Interval interval);
```

Boolean flag style method often result in pretty confusing function signatures, because you are not always sure of what the boolean is supposed to mean.  Especially if you don't actually have the code.  If we can instead move those semantics into a completely different method name, it becomes easier for users to know what is going on.

Also, :+1: on adding javadoc.

File: server/src/main/java/io/druid/client/CoordinatorServerView.java (hunk 4 above)
Comment (cheddar) [inline server/src/main/java/io/druid/client/CoordinatorServerView.java]: After looking at the rest of the code, I understand why you are doing this.  It makes sense, so consider this comment rescinded.

File: server/src/main/java/io/druid/client/coordinator/CoordinatorClient.java (hunk 5 above)
Comment (xvrl) [inline server/src/main/java/io/druid/client/coordinator/CoordinatorClient.java]: I agree we should separate external from internal service discovery. However, is there any reason we can't have two `ServerDiscoverySelector`s, one for internal, and one for external purposes? It seems easier to work with and less error-prone than handling the announcement manually for pure discovery, which is part of the reason for why it was used in `RemoteTaskActionClient` instead of relying on announcer.

File: server/src/main/java/io/druid/client/TimelineInventoryView.java [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package io.druid.client;
+
+import io.druid.query.DataSource;
+import io.druid.timeline.TimelineLookup;
+
+public interface TimelineInventoryView extends InventoryView
```
Comment (cheddar) [inline server/src/main/java/io/druid/client/TimelineInventoryView.java]: From what I can tell, there is only one instance of implementing `TimelineInventoryView` and it is the `CoordinatorServerView`.  Even in `DatasourcesResourceTest` you are mocking `CoordinatorServerView` instead of implementing this interface.  There are two reasons to add an interface

1) You have multiple implementations and want polymorphic dispatch (in tests or in normal code)
2) You expect people to extend the interface and add their own implementations

Given that there is only one implementation of the interface and it doesn't seem to be necessary for the tests, I don't think (1) applies.  I also do not expect (2) to be an issue, this is an internal implementation rather than something I expect someone to extend.

If you agree with my assessment above, then please remove this interface entirely and just depend on the concrete `CoordinatorServerView`.  If not, please let me know what I'm missing.

File: server/src/main/java/io/druid/client/coordinator/CoordinatorClient.java (hunk 10 above)
Comment (xvrl) [inline server/src/main/java/io/druid/client/coordinator/CoordinatorClient.java]: We seem to be conflicted about this. The `/druid/coordinator/v1/datasources` endpoint uses `/` for `DELETE /{dataSourceName}?interval=...` **query parameters**, but uses `_` for `GET /{dataSourceName}/intervals/{interval}` **path parameters**

File: common/src/main/java/io/druid/timeline/TimelineLookup.java (hunk 1 above)
Comment (nishantmonu51) [inline common/src/main/java/io/druid/timeline/TimelineLookup.java]: fixed.

File: indexing-service/src/main/java/io/druid/indexing/common/TaskToolboxFactory.java (hunk 2 above)
Comment (nishantmonu51) [inline indexing-service/src/main/java/io/druid/indexing/common/TaskToolboxFactory.java]: fixed.

File: server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java (hunk 3 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java]: yeah, the logic is same.

File: server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java (hunk 6 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java]: fixed.

File: server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java (hunk 7 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java]: fixed.

File: server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java (hunk 8 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java]: done.

File: server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java (hunk 9 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java]: changes in this file include - 
- replacing serverView with handoffNotifier 
- it registers callback with handoffNotifier when pushing a segment
- moved the logic for determining when handoff is complete from plumber to the handoff notifier.

File: server/src/main/java/io/druid/client/TimelineInventoryView.java (hunk 11 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/client/TimelineInventoryView.java]: removed unnecessary interface.

File: server/src/main/java/io/druid/client/coordinator/CoordinatorClient.java (hunk 10 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/client/coordinator/CoordinatorClient.java]: I agree we should be consistent with the way interval is handled, 
I think we can use _ for all GET, POST and DELETE endpoints as a consistent way to handle intervals. we might also want to keep the old behavior for delete end-points for backwards compatibility. however i would like to do it in a separate PR.
@xvrl @cheddar does this sound good to you ?

**Comment by xvrl** [comment]
:+1: looks good to me.

If we change the way coordinator discovery is done, we should also change the way overlord discovery in `RemoteTaskActionClient` is done, since they are effectively the same. In order to avoid scope creep of this PR, I would be in favor of doing this change in a separate one. I'll file an issue for this.

File: docs/content/configuration/index.md [hunk 12]
```diff
@@ -257,6 +257,16 @@ This config is used to find the [Indexing Service](../design/indexing-service.ht
...
 |`druid.selectors.indexing.serviceName`|The druid.service name of the indexing service Overlord node. To start the Overlord with a different name, set it with this property. |druid/overlord|
 
+
+### Coordinator Discovery
+
+This config is used to find the [Coordinator](../design/coordinator.html) using Curator service discovery. This config is used by the realtime indexing nodes to get information about the segments loaded in the cluster.
+
+|Property|Description|Default|
+|--------|-----------|-------|
+|`druid.selectors.coordinator.serviceName`|The druid.service name of the coordinator node. To start the Coordinator with a different name, set it with this property. |druid/coordinator|
```
Comment (gianm) [inline docs/content/configuration/index.md]: This should be set to "coordinator" in the example common.runtime.properties, since the example coordinator properties set the druid.service to "coordinator". Many people start off their clusters by copying the example configs.

I wonder if we should also just change the defaults in the code to match those…

Comment (xvrl) [inline docs/content/configuration/index.md]: I think the configs were created before we had defaults. I'd rather just change the example configs to match the defaults, or leave it out of the example configs entirely, since it's not necessary.

File: server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java [hunk 13]
```diff
@@ -597,6 +584,17 @@ public void doRun()
...
+        new SegmentDescriptor(sink.getInterval(), sink.getVersion(), config.getShardSpec().getPartitionNum()),
+        mergeExecutor, new Runnable()
+        {
+          @Override
+          public void run()
+          {
+            abandonSegment(sink.getInterval().getStartMillis(), sink);
```
Comment (gianm) [inline server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java]: It's possible that the sink is no longer in `sinks` at this point, so this runnable should check for that.

File: server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java [hunk 14]
```diff
@@ -735,97 +734,101 @@ public int compare(File o1, File o2)
           }
       );
       boolean isCorrupted = false;
-      try {
```
Comment (gianm) [inline server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java]: Why remove this try/catch?

File: docs/content/configuration/index.md (hunk 12 above)
Comment (gianm) [inline docs/content/configuration/index.md]: The default overlord service name in tranquility is "overlord" to match the examples, so if we do change the examples we should change that too. All three of those things should match though (druid defaults, tranquility defaults, druid example configs)

File: server/src/main/java/io/druid/client/CoordinatorServerView.java [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+  @Inject
+  public CoordinatorServerView(
+      ServerInventoryView baseView
+  )
+  {
+    this.baseView = baseView;
+    this.selectors = Maps.newHashMap();
+    this.timelines = Maps.newHashMap();
+
+    ExecutorService exec = Execs.singleThreaded("OverlordServerView-%s");
```
Comment (gianm) [inline server/src/main/java/io/druid/client/CoordinatorServerView.java]: CoordinatorServerView

File: server/src/main/java/io/druid/client/CoordinatorServerView.java [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+/**
+ * ServerView of coordinator for the state of segments being loaded in the cluster.
+ */
+public class CoordinatorServerView implements InventoryView
+{
+  private static final Logger log = new Logger(CoordinatorServerView.class);
+
+  private final Object lock = new Object();
+
+  private final Map<String, Set<DruidServerMetadata>> selectors;
```
Comment (gianm) [inline server/src/main/java/io/druid/client/CoordinatorServerView.java]: these are servers, not selectors; rename?

File: server/src/main/java/io/druid/client/CoordinatorServerView.java [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+      servers.remove(server);
+      if (servers.isEmpty()) {
+        VersionedIntervalTimeline<String, SegmentLoadInfo> timeline = timelines.get(segment.getDataSource());
+        selectors.remove(segmentId);
+
+        final PartitionChunk<SegmentLoadInfo> removedPartition = timeline.remove(
+            segment.getInterval(), segment.getVersion(), segment.getShardSpec().createChunk(
+                new SegmentLoadInfo(
+                    segment,
+                    Sets
```
Comment (gianm) [inline server/src/main/java/io/druid/client/CoordinatorServerView.java]: `ImmutableSet.of`?

Comment (gianm) [inline server/src/main/java/io/druid/client/CoordinatorServerView.java]: or are you actually modifying this?

Comment (gianm) [inline server/src/main/java/io/druid/client/CoordinatorServerView.java]: ok, I see, you are modifying it. Those modifications should be made thread safe then, as the SegmentLoadInfos are getting escaped outside of this object.

File: server/src/main/java/io/druid/client/SegmentLoadInfo.java [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+import io.druid.server.coordination.DruidServerMetadata;
+import io.druid.timeline.DataSegment;
+
+import java.util.List;
+import java.util.Set;
+
+public class SegmentLoadInfo
+{
+  private final DataSegment segment;
+  private final Set<DruidServerMetadata> servers;
```
Comment (gianm) [inline server/src/main/java/io/druid/client/SegmentLoadInfo.java]: It looks like this is potentially going to be modified by the CoordinatorServerView while a caller might have a reference to the load info, so it should be a `ConcurrentSet`.

File: server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+    this.pollDurationMillis = config.getPollDuration().getMillis();
+  }
+
+  @Override
+  public void registerSegmentHandoffCallback(
+      SegmentDescriptor descriptor, Executor exec, Runnable handOffRunnable
+  )
+  {
+    log.info("Adding SegmentHandoffCallback for dataSource[%s] Segment[%s]", dataSource, descriptor);
+    handOffCallbacks.put(descriptor, new Pair<>(exec, handOffRunnable));
```
Comment (gianm) [inline server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java]: `putIfAbsent`, return false if one's already there? The current thing makes it pretty easy to accidentally overwrite a callback.

File: server/src/main/java/io/druid/server/http/DatasourcesResource.java [hunk 20]
```diff
@@ -554,4 +563,55 @@ public DruidDataSource apply(DruidServer input)
...
+      @QueryParam("partial") final boolean partial
+  )
+  {
+    TimelineLookup<String, SegmentLoadInfo> timeline = serverInventoryView.getTimeline(
+        new TableDataSource(dataSourceName)
+    );
+    final Interval theInterval = new Interval(interval.replace("_", "/"));
+    if (timeline == null) {
+      log.debug("No timeline found for datasource[%s]", dataSourceName);
+      return Response.noContent().build();
```
Comment (gianm) [inline server/src/main/java/io/druid/server/http/DatasourcesResource.java]: Why not return an empty array? It's ok for a datasource/interval pair to not have any served segments.

File: server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+                      descriptor.getInterval(),
+                      true
+                  );
+
+                  if (isHandOffComplete(loadedSegments, entry.getKey())) {
+                    log.info("Segment Handoff complete for dataSource[%s] Segment[%s]", dataSource, descriptor);
+                    entry.getValue().lhs.execute(entry.getValue().rhs);
+                    itr.remove();
+                  } else {
+                    log.info(
```
Comment (gianm) [inline server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java]: Print these all at the end of the loop?

Like: `Still waiting for handoff of segments: blah`

**Comment by gianm** [comment]
The approach generally seems good although there are a couple of comments in the diff that I think are important.

Other than those- shall we also remove the FilteredServerView interface and `registerSegmentCallback(Executor, ServerView.SegmentCallback, Predicate<DataSegment>)` methods from the various inventory views? This was the only user of them and so nothing uses that ability anymore.

File: server/src/main/java/io/druid/client/coordinator/CoordinatorClient.java (hunk 5 above)
Comment (gianm) [inline server/src/main/java/io/druid/client/coordinator/CoordinatorClient.java]: IMO a follow-on PR is OK, maybe we should file an issue to remember that.

**Comment by xvrl** [comment]
Agree with @gianm we can remove `FilteredServerView` which was only meant to reduce memory pressure on realtime node serverview. I'm fine doing this here or in a separate PR/commit. Regarding my earlier comment, I filed issue #2040.

File: docs/content/configuration/index.md (hunk 12 above)
Comment (nishantmonu51) [inline docs/content/configuration/index.md]: https://github.com/druid-io/druid/issues/2046 to track

File: server/src/main/java/io/druid/client/CoordinatorServerView.java (hunk 16 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/client/CoordinatorServerView.java]: done.

File: server/src/main/java/io/druid/client/CoordinatorServerView.java (hunk 15 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/client/CoordinatorServerView.java]: fixed.

File: server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java (hunk 19 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java]: done.

File: server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java (hunk 21 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java]: done.

File: server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java (hunk 13 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java]: fixed.

File: server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java (hunk 14 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java]: this was catching the IOException which could only come from announcer inside this moved this inside the addSink method, other IOEx inside this try/catch were already handled by the inner catch blocks.

File: server/src/main/java/io/druid/client/SegmentLoadInfo.java (hunk 18 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/client/SegmentLoadInfo.java]: fixed. CoordinatorServerView  now passes a concurrentHashSet.

File: server/src/main/java/io/druid/client/CoordinatorServerView.java (hunk 17 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/client/CoordinatorServerView.java]: fixed.

**Comment by nishantmonu51** [comment]
@xvrl @gianm , handled review comments and removed FilteredServerView in a separate commit. 
please have a look again.

File: docs/content/configuration/index.md (hunk 12 above)
Comment (nishantmonu51) [inline docs/content/configuration/index.md]: https://github.com/druid-io/druid/pull/2048

**Comment by pjain1** [comment]
Standalone RealtimeNode fails to start with Guice provision errors -
```
1) Error injecting constructor, java.lang.RuntimeException: com.fasterxml.jackson.databind.JsonMappingException: Guice configuration errors:

1) No implementation for io.druid.segment.realtime.plumber.SegmentHandoffNotifier was bound.
  while locating io.druid.segment.realtime.plumber.SegmentHandoffNotifier

1 error (through reference chain: java.util.ArrayList[0])
  at io.druid.guice.FireDepartmentsProvider.<init>(FireDepartmentsProvider.java:41)
  while locating io.druid.guice.FireDepartmentsProvider
  at io.druid.guice.RealtimeModule.configure(RealtimeModule.java:88)
  while locating java.util.List<io.druid.segment.realtime.FireDepartment>
    for parameter 0 at io.druid.segment.realtime.RealtimeManager.<init>(RealtimeManager.java:80)
  while locating io.druid.segment.realtime.RealtimeManager
  at io.druid.guice.RealtimeModule.configure(RealtimeModule.java:102)
  while locating io.druid.query.QuerySegmentWalker
    for parameter 3 at io.druid.server.QueryResource.<init>(QueryResource.java:90)
  while locating io.druid.server.QueryResource
```
I guess the problem is that in case of standalone node, `RealtimePlumberSchool` object is directly initialized by Guice when `FireDepartment` object is created and it is expecting binding for `SegmentHandoffNotifier`

**Comment by nishantmonu51** [comment]
@pjain1 thanks for checking it out, looking into it.

File: services/src/main/java/io/druid/cli/CliCoordinator.java [hunk 22]
```diff
@@ -117,6 +118,8 @@ public void configure(Binder binder)
                   .in(ManageLifecycle.class);
 
             binder.bind(IndexingServiceClient.class).in(LazySingleton.class);
+            binder.bind(CoordinatorServerView.class).in(LazySingleton.class);
+            binder.bind(CoordinatorServerView.class).in(LazySingleton.class);
```
Comment (pjain1) [inline services/src/main/java/io/druid/cli/CliCoordinator.java]: binding twice ?

File: server/src/main/java/io/druid/segment/realtime/plumber/SegmentHandoffNotifier.java [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied. See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package io.druid.segment.realtime.plumber;
+
+import io.druid.query.DataSource;
```
Comment (pjain1) [inline server/src/main/java/io/druid/segment/realtime/plumber/SegmentHandoffNotifier.java]: unused import ?

**Comment by nishantmonu51** [comment]
@pjain1: the realtime node got broken when i refactored things during handling review comments, fixed now.

**Comment by xvrl** [comment]
@gianm @pjain1 any more comments here?

**Comment by pjain1** [comment]
I am :+1:

File: server/src/main/java/io/druid/client/SegmentLoadInfo.java (hunk 18 above)
Comment (gianm) [inline server/src/main/java/io/druid/client/SegmentLoadInfo.java]: Can you change this type to a `ConcurrentSet` to emphasize that it must be concurrent?

File: server/src/main/java/io/druid/client/CoordinatorServerView.java [hunk 24]
```diff
@@ -0,0 +1,213 @@
...
+    }
+  }
+
+  private void serverAddedSegment(final DruidServerMetadata server, final DataSegment segment)
+  {
+    String segmentId = segment.getIdentifier();
+    synchronized (lock) {
+      log.debug("Adding segment[%s] for server[%s]", segment, server);
+
+      Set<DruidServerMetadata> servers = this.servers.get(segmentId);
```
Comment (gianm) [inline server/src/main/java/io/druid/client/CoordinatorServerView.java]: Can you please rename this, it's confusing to have servers here and also servers as a class member.

File: server/src/main/java/io/druid/client/CoordinatorServerView.java [hunk 25]
```diff
@@ -0,0 +1,213 @@
...
+      if (servers.isEmpty()) {
+        VersionedIntervalTimeline<String, SegmentLoadInfo> timeline = timelines.get(segment.getDataSource());
+        this.servers.remove(segmentId);
+
+        final PartitionChunk<SegmentLoadInfo> removedPartition = timeline.remove(
+            segment.getInterval(), segment.getVersion(), segment.getShardSpec().createChunk(
+                new SegmentLoadInfo(
+                    segment,
+                    Sets
+                        .<DruidServerMetadata>newHashSet()
```
Comment (gianm) [inline server/src/main/java/io/druid/client/CoordinatorServerView.java]: this should be a concurrent set

File: server/src/main/java/io/druid/client/SegmentLoadInfo.java [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+
+  @JsonCreator
+  public SegmentLoadInfo(
```
Comment (gianm) [inline server/src/main/java/io/druid/client/SegmentLoadInfo.java]: When is this going to be serialized / deserialized?

Comment (gianm) [inline server/src/main/java/io/druid/client/SegmentLoadInfo.java]: I'm wondering if we can force the set to always be concurrent here to avoid issues caused by accidentally getting non-concurrent sets in here

Comment (nishantmonu51) [inline server/src/main/java/io/druid/client/SegmentLoadInfo.java]: this is serialized when the realtime task polls about the status of segments loaded via http end point. java doesn't have any ConcurrentSet api which can be used to enforce it. 
Right now only coordinatorServerView creates that object. 
Do you have any suggestions on how we can enforce it ?

Comment (gianm) [inline server/src/main/java/io/druid/client/SegmentLoadInfo.java]: ah, good point. ConcurrentHashSet, then?

Comment (nishantmonu51) [inline server/src/main/java/io/druid/client/SegmentLoadInfo.java]: working on it by having an internal object and external immutable serializable object.

Comment (gianm) [inline server/src/main/java/io/druid/client/SegmentLoadInfo.java]: that sounds good, especially because none of those interfaces I wanted to exist actually existed :)

Comment (xvrl) [inline server/src/main/java/io/druid/client/SegmentLoadInfo.java]: sounds like this might be better as a separate PR?

Comment (gianm) [inline server/src/main/java/io/druid/client/SegmentLoadInfo.java]: @xvrl I think it makes sense in this one, since this is a new class and it is important to get its thread safety right.

Comment (nishantmonu51) [inline server/src/main/java/io/druid/client/SegmentLoadInfo.java]: done. added ImmutableSegmentLoadInfo

**Comment by gianm** [comment]
:+1: after travis

**Comment by nishantmonu51** [comment]
bouncing for travis.

**Comment by nishantmonu51** [comment]
Also added graph showing the improvement on zookeeper load  with ~500 realtime index tasks running in PR description.

**Comment by gianm** [comment]
I think this patch makes the rolling update process not work, since it suggests doing coordinators after indexing nodes, but we'd need the coordinator endpoint up first.

**Comment by gianm** [comment]
added a "release notes" label

**Comment by gianm** [comment]
We also need to add to the release notes that you need to make sure your `druid.selectors.coordinator.serviceName` is set properly. Otherwise realtime indexing will stop working after updating because the handoff notifier won't be linked up.

**Comment by xvrl** [comment]
@gianm oddly this PR is not even mentioned at all in the release notes

**Comment by gianm** [comment]
Probably it was added after the first RC and missed. We should double check that all the stuff after that actually made it into the notes.

**Comment by gianm** [comment]
@xvrl will take a look now

**Comment by gianm** [comment]
@xvrl updated the release notes & added a few other missing things

**Comment by xvrl** [comment]
@gianm thx I was updating the release notes but you beat me to it.

**Comment by rasahner** [comment]
I just realized that the new endpoint isn't documented at http://druid.io/docs/latest/design/coordinator.html.  Should it be?  Or, are there some types of endpoints that are deliberately not documented at druid.io/docs and this is one of them?

**Comment by fjy** [comment]
@rasahner This should be documented. Can you submit a PR?

**Comment by rasahner** [comment]
see https://github.com/druid-io/druid/pull/2238

**Comment by fjy** [comment]
@nishantmonu51 @gianm @pjain1 @xvrl @cheddar 

None of the examples were updated to reflect that serviceName now talks to coordinator. None of the logs were updated either. Handoff is broken in the examples and the logs that Druid currently display also don't make sense.

**Comment by fjy** [comment]
Also, does anything that requires the coordinator to talk to the overlord still work? Kill tasks, merge tasks, etc?
